src/dmd/toir.d

```diff
@@ -361,311 +361,93 @@ elem *setEthis(const ref Loc loc, IRState *irs, elem *ey, AggregateDeclaration a
+
+    id2 = md.id;
+
+    if (md.packages.dim != 1)
+        return op;
+
+    id1 = (*md.packages)[0];
+
+    if (id1 == Id.std && id2 == Id.math)
     {
-        for (size_t i = 0; i < std_namearray.length - 1; i++)
-        {
-            if (strcmp(std_namearray[i], std_namearray[i + 1]) >= 0)
-            {
-                printf("std_namearray[%ld] = '%s'\n", cast(long)i, std_namearray[i]);
-                assert(0);
-            }
-        }
-        for (size_t i = 0; i < std_namearray64.length - 1; i++)
+        Lmath:
+        if (id3 == Id.cos)    op = OPcos;
+        if (id3 == Id.sin)    op = OPsin;
+        if (id3 == Id.fabs)   op = OPabs;
+        if (id3 == Id.rint)   op = OPrint;
+        if (id3 == Id._sqrt)  op = OPsqrt;
+        if (id3 == Id.yl2x)   op = OPyl2x;
+        if (id3 == Id.ldexp)  op = OPscale;
+        if (id3 == Id.rndtol) op = OPrndtol;
+        if (id3 == Id.yl2xp1) op = OPyl2xp1;
+    }
+    else if (id1 == Id.core)
+    {
+        if (id2 == Id.math)
         {
-            if (strcmp(std_namearray64[i], std_namearray64[i + 1]) >= 0)
-            {
-                printf("std_namearray64[%ld] = '%s'\n", cast(long)i, std_namearray64[i]);
-                assert(0);
-            }
+            goto Lmath;
         }
-        for (size_t i = 0; i < core_namearray.length - 1; i++)
+        else if (id2 == Id.simd)
         {
-            //printf("test1 %s %s %d\n", core_namearray[i], core_namearray[i + 1], strcmp(core_namearray[i], core_namearray[i + 1]));
-            if (strcmp(core_namearray[i], core_namearray[i + 1]) >= 0)
-            {
-                printf("core_namearray[%ld] = '%s'\n", cast(long)i, core_namearray[i]);
-                assert(0);
-            }
+            // __prefetch
+            if (id3 == Id.__prefetch) op = OPprefetch;
+            // __simd_sto, __simd, __simd_ib
+            else                      op = OPvector;
         }
-        for (size_t i = 0; i < core_namearray64.length - 1; i++)
+        else if (id2 == Id.bitop)
         {
-            if (strcmp(core_namearray64[i], core_namearray64[i + 1]) >= 0)
-            {
-                printf("core_namearray64[%ld] = '%s'\n", cast(long)i, core_namearray64[i]);
-                assert(0);
-            }
+            if (id3 == Id.volatileLoad)  op = OPind;
+            if (id3 == Id.volatileStore) op = OPeq;
+
+            if (id3 == Id.bsf) op = OPbsf;
+            if (id3 == Id.bsr) op = OPbsr;
+            if (id3 == Id.btc) op = OPbtc;
+            if (id3 == Id.btr) op = OPbtr;
+            if (id3 == Id.bts) op = OPbts;
+
+            if (id3 == Id.inp)  op = OPinp;
+            if (id3 == Id.inpl) op = OPinp;
+            if (id3 == Id.inpw) op = OPinp;
+
+            if (id3 == Id.outp)  op = OPoutp;
+            if (id3 == Id.outpl) op = OPoutp;
+            if (id3 == Id.outpw) op = OPoutp;
+
+            if (id3 == Id.bswap)   op = OPbswap;
+            if (id3 == Id._popcnt) op = OPpopcnt;
```

Comment:
It would be better to check the identifier length first before the pointer comparison.
i.e: (pseudo-code)
```
switch (id3.length)
{
  case 3:
    if (id3 == Id.bsf)
      op = OPbsf;
    else if (...)
      // ...
    break;
  case ...:
    // ...
}
```

Reply:
true. Will do.

Reply:
Hmm, LDC/GDC should be able to vectorise it pretty well without the switch as a masked gather , DMD obviously won't do that. Switching on the length requires loading it, its probably cold: I don't think it would buy all that much over the if chains because they are pretty short.

Reply:
Well, there were once only half a dozen or so intrinsics in the past, that list has grown considerably since then, and it will only continue to grow.